compiler-rt: Provide `__cpu_indicator_init`, `__cpu_model` and `__cpu_features2`

lacc97 opened the pull request:
Fixes https://github.com/ziglang/zig/issues/18074.

Adds `__cpu_indicator_init`, `__cpu_model` and `__cpu_features2` symbols to compiler-rt for x86/x86_64. These match what LLVM expects and generates when building multiversioned functions (e.g. functions tagged with attributes `target` or `target_clones`). The actual CPU detection reuses the logic under std/zig/system/x86.zig, and transforms it into the values that LLVM-generated code expects. These values are auto-generated using a new tool (tools/update_x86_cpu_model_enums.zig) which parses the relevant file in the LLVM codebase and generates Zig enums out of it.

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,428 @@
+    set[f_value / 32] |= (@as(u32, 1) << @as(u5, @intCast(f_value % 32)));
+}
+
+// These values match the values in LLVM.
```
Can we make checks for this to detect divergence from LLVM? Maybe via filtering the target feature set?

matu3ba replied:
I was wondering, if something like https://github.com/ziglang/zig/blob/8a8fd47d217eadb8a45ec6b9325acfe1d42704d7/src/zig_llvm.cpp#L1174, ie as being part of https://github.com/ziglang/zig/blob/8a8fd47d217eadb8a45ec6b9325acfe1d42704d7/tools/update_cpu_features.zig would be feasible to keep things synchronized with upstream.

I don't think this is a requirement to merge, just something nice to have for maintenance.

lacc97 replied:
Ohhh I see that's how Zig generates the `std.Target.x86` enums. Unfortunately the compiler-rt feature enums seem to be completely separate from the features defined in `llvm/lib/Target/X86/X86.td` which is why the upstream implementation of cpu_model looks kind of sloppy with different feature names or unused features.

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,428 @@
+var cpu: Model = .{};
+var cpu_extra_features: [3]u32 = [_]u32{0} ** 3;
+
+// TODO: constructor attribute?
```
@kubkon Does Zig intend to use `constructor`/`destructor` (see for example here https://stackoverflow.com/questions/6589772/gcc-functions-with-constructor-attribute-are-not-being-linked) or how do you think this should be solved?

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,428 @@
+    avx512pf,
+    avx512vbmi,
+    avx512ifma,
+    avx5124vnniw, // TODO: detect this in std/zig/system/x86.zig
```
The general approach is to eliminate TODOs to make the context explicit to readers and simplify external contributions etc.
If you don't want to make this in this PR or a follow-up, then please make a very brief text what would be needed so another person can dive in quickly.

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,428 @@
+    avx512bf16,
+    avx512vp2intersect,
+
+    cx16 = 46, // NOTE: named `cmpxchg16b` in LLVM source
```
1. Andrew thinks that every comment is a note. 
2. Did you rename them or has this been done in the correlated libstd code? I'm wondering, if LLVM changed names between versions or why the Zig names are differently here from the LLVM ones.

lacc97 replied:
And I'll remove the `NOTE: ` bit.

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,428 @@
+};
+
+comptime {
+    // Check that every name in Feature has a corresponding name  in std.Target.x86.Feature.
```
Can we please also check vice versa so things remain synced?

lacc97 replied:
I went from `Feature` -> `std.Target.x86.Feature` because there are a lot more values in `std.Target.x86.Feature` which we can already detect in `std.zig.x86`. Since the only features we have to care about in this file are the ones LLVM understands I decided to only check that every enum name in `Feature` has a corresponding enum name in `std.Target.x86.Feature` (with specific exceptions) and not the other way around.

matu3ba commented on lib/std/zig/system/x86.zig:
```diff
@@ -31,13 +31,52 @@ pub fn detectNativeCpuAndFeatures(arch: Target.Cpu.Arch, os: Target.Os, cross_ta
+        }
+    }
+
+    // Add the CPU model's feature set into the working set, but then
```
nit: "working set" is not a good name. Can you/anyone think of a better phrase? Maybe use variable name here.

Also, maybe this should justify why this is needed and okay and does not break user assumptions on the target features: The naive reader would read that target baseline would be overwritten by current feature set of the host cpu.

lacc97 replied:
This is already existing code I moved around and I'm not exactly sure why we do this.

matu3ba reviewed (commented):
That was fast. Love it.

Mostly some ideas around some explanations that things are correct and how to keep them correct and maintainable (for future readers/adjustments).

I did not have a deep look into the details yet, only in the overall organization and explanations.

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,428 @@
+        .{ "x86_64_v3", {} },
+        .{ "x86_64_v4", {} },
+
+        // Unused in LLVM
```
Could you maybe make it clear then, what is part of LLVM (and unused there) and what not? This leaves the impression of code smell in Zig's repo.

matu3ba commented:
> This does mean that I'll basically have to rewrite this merge request haha

Sorry about that, I was expecting LLVM would be more structured.

lacc97 commented:
> > This does mean that I'll basically have to rewrite this merge request haha
> 
> Sorry about that, I was expecting LLVM would be more structured.

No worries :)

I think it's ready now. As I mentioned yesterday, the enums are no longer manually hardcoded and instead are generated from the LLVM source, specifically llvm/include/llvm/TargetParser/X86TargetParser.def. The tool that generates these enums lives in tools/update_x86_cpu_model_enums.zig, which calls into `zig cc` as a C preprocessor with the appropriate defines to extract the enum names and the order in which they are defined so that they match what LLVM expects. Since these enums might change from LLVM version to version (hopefully on major versions only), the tool would need to be run after an LLVM version update.

I also changed the test a bit to also check for `__cpu_features2` symbol which is used for the features with bit index greater than 32 (which means they no longer fit in `__cpu_model`).

matu3ba commented on test/standalone/x86_cpu_model/main.c:
```diff
@@ -0,0 +1,19 @@
+}
+int __attribute__ ((__target__("avx512vnni"))) foo(void) {
+    return 3;
+}
```
nit: standard compliant must include a newline at end of file or might introduce ub, see https://gcc.gnu.org/legacy-ml/gcc/2001-07/msg01120.html

lacc97 replied:
Oh wow I did not know that could happen.

https://github.com/ziglang/zig/pull/18193/commits/6d50d8715a46e7ea30aa59a07261067f86390333

matu3ba commented on tools/update_x86_cpu_model_enums.zig:
```diff
@@ -0,0 +1,196 @@
+const std = @import("std");
```
make sure that this tool is ci tested in test/standalone.zig:63.

lacc97 replied:
https://github.com/ziglang/zig/pull/18193/commits/2841c9e61916309884093544f749ece6afb67d5f

matu3ba reviewed (commented):
Looking good now. Just some more nits.

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,281 @@
+    // Sanity check that every name in Feature has a corresponding name in std.Target.x86.Feature.
+
+    const ignored = std.ComptimeStringMap(void, .{
+        // These are not present in std/target/x86.zig because LLVM does
```
On master this is not entirely correct, they are unimplemented stubs:
```
// FIXME: These two aren't really implemented and just exist in the feature
// list for __builtin_cpu_supports. So omit their dependencies.
constexpr FeatureBitset ImpliedFeaturesAVX5124FMAPS = {};
constexpr FeatureBitset ImpliedFeaturesAVX5124VNNIW = {};
```

matu3ba commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,281 @@
+    if (missing) @compileError("missing enum values in std.Target.x86.Feature (see compile log)");
+}
+
+// This function should be reviewed after any update to the autogenerated
```
Better make this dry and note the functions to review in the file level comment-

matu3ba reviewed (changes requested):
some more nits. Will look again once this works in CI.

matu3ba reviewed (approved):
Looking good. The person merging this should try out the tooling file locally to confirm that this works.

lacc97 commented:
Rebased on latest master.

lacc97 commented:
@andrewrk I rebased on latest master. I also added a couple of commits so that the tool works on LLVM 18 as well (it might still need more work on the compiler-rt side for LLVM 18 but at least the tool should generate the right enum values).

andrewrk commented:
I'm sorry, I didn't review this in time, and now it has bitrotted. Furthermore, so many pull requests have stacked up that I can't keep up and I am therefore declaring Pull Request Bankruptcy and closing old PRs that now have conflicts with master branch.

If you want to reroll, you are by all means welcome to revisit this changeset with respect to the current state of master branch, and there's a decent chance your patch will be reviewed the second time around.

Either way, I'm closing this now, otherwise the PR queue will continue to grow indefinitely.

lacc97 commented:
No worries, I made a new PR here https://github.com/ziglang/zig/pull/20081.
